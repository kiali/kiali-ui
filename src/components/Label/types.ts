export interface Labels {
  name: string;
  value: string;
}
