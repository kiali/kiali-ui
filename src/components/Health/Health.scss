svg:not(:root).icon-failure,
svg:not(:root).icon-degraded,
svg:not(:root).icon-healthy,
svg:not(:root).icon-na {
  overflow: visible;
}
.tippy-popper[x-placement^='left'] .pf-tooltip-theme .tippy-arrow {
  // This is the old style of health tooltip with white background
  // I will keep it here instead to remove completely the style in case we need tweaks in the future
  // background-color: #fff;
  // border-color: black;
  // border-width: 1px;
  // border-style: solid;
}
.health_indicator .pf-c-tooltip__content {
  // This is the old style of health tooltip with white background
  // I will keep it here instead to remove completely the style in case we need tweaks in the future
  // background-color: #fff;
  // border-style: solid;
  // border-color: black;
  border-width: 1px;
  text-align: left;
}
.health_indicator .pf-c-content ul {
  margin-bottom: var(--pf-c-content--ul--MarginTop);
  margin-top: 0;
  color: #fff;
  --pf-global--Color--100: var(--pf-global--Color--light-100);
}

.health_indicator .pf-c-content small {
  font-size: 70%;
  color: var(--pf-global--Color--light-100);
}

.health_indicator .pf-m-gutter {
  --pf-l-split--m-gutter--MarginRight: 0.25rem;
  --pf-l-split--m-gutter--md--MarginRight: 0.25rem;
}

#health .pf-m-gutter {
  --pf-l-split--m-gutter--MarginRight: 0.25rem;
  --pf-l-split--m-gutter--md--MarginRight: 0.25rem;
}
