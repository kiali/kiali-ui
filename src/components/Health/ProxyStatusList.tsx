import * as React from 'react';
import { ProxyStatus } from '../../types/Health';
import { Flex, FlexItem, Split, SplitItem, Text, TextVariants } from '@patternfly/react-core';
import { style } from 'typestyle';

type Props = {
  workloadName: string;
  statuses: ProxyStatus[];
};

const smallStyle = style({ fontSize: '70%' });
const workloadNameStyle = style({ minWidth: '6rem', overflowWrap: 'break-word' });

const proxyStatusSort = (a: ProxyStatus, b: ProxyStatus): number => {
  return a.component > b.component ? 1 : -1;
};

class ProxyStatusList extends React.Component<Props> {
  sortedStatus = () => {
    return this.props.statuses.sort(proxyStatusSort);
  };

  statusList = () => {
    const statusItems = this.sortedStatus().map((value: ProxyStatus, i: number) => {
      return (
        <FlexItem key={'proxy-status-' + i}>
          <Text component={TextVariants.small} className={smallStyle}>
            {value.component + ': ' + value.status}
          </Text>
        </FlexItem>
      );
    });

    return <Flex style={{ flex: 'inlineFlex' }}> {statusItems} </Flex>;
  };

  render() {
    if (this.props.statuses.length > 0) {
      return (
        <Split gutter={'sm'}>
          <SplitItem className={workloadNameStyle}>{this.props.workloadName}</SplitItem>
          <SplitItem>{this.statusList()}</SplitItem>
        </Split>
      );
    } else {
      return [];
    }
  }
}

export default ProxyStatusList;
