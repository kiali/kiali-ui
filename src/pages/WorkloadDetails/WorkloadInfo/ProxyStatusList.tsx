import * as React from 'react';
import { isProxyStatusComponentSynced, isProxyStatusSynced, ProxyStatus } from '../../../types/Health';
import { Stack, StackItem } from '@patternfly/react-core';
import { style } from 'typestyle';
import { PfColors } from '../../../components/Pf/PfColors';

type Props = {
  status?: ProxyStatus;
};

const smallStyle = style({ fontSize: '70%', color: PfColors.White });
const colorStyle = style({ fontSize: '1.1rem', color: PfColors.White });

class ProxyStatusList extends React.Component<Props> {
  statusList = () => {
    if (!this.props.status) {
      return [];
    }

    return [
      { c: 'CDS', s: this.props.status.CDS },
      { c: 'EDS', s: this.props.status.EDS },
      { c: 'LDS', s: this.props.status.LDS },
      { c: 'RDS', s: this.props.status.RDS }
    ].map((value: { c: string; s: string }, i: number) => {
      if (!isProxyStatusComponentSynced(value.s)) {
        const status = value.s ? value.s : '-';
        return (
          <StackItem key={'proxy-status-' + i} className={smallStyle}>
            {value.c + ': ' + status}
          </StackItem>
        );
      } else {
        return null;
      }
    });
  };

  render() {
    if (this.props.status && !isProxyStatusSynced(this.props.status)) {
      return (
        <Stack>
          <StackItem className={colorStyle}>Istio Proxy Status</StackItem>
          {this.statusList()}
        </Stack>
      );
    } else {
      return null;
    }
  }
}

export default ProxyStatusList;
