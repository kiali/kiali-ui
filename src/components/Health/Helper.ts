import * as React from 'react';
import { ProxyStatus, Status } from 'types/Health';
import ProxyStatusList from './ProxyStatusList';

type Size = 'sm' | 'md' | 'lg' | 'xl';

export const createIcon = (status: Status, size?: Size) => {
  return React.createElement(status.icon, { color: status.color, size: size, className: status.class });
};

export const createProxyStatusList = (workloadName: string, statusList: ProxyStatus[]) => {
  return React.createElement(ProxyStatusList, { workloadName: workloadName, statuses: statusList });
};
