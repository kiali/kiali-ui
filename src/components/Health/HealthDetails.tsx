import * as React from 'react';
import * as H from '../../types/Health';
import { createIcon } from './Helper';
import { InfoAltIcon } from '@patternfly/react-icons';
import './Health.css';
import { PfColors } from '../Pf/PfColors';

interface Props {
  health: H.Health;
  tooltip?: boolean;
}

export class HealthDetails extends React.PureComponent<Props, {}> {
  renderErrorRate = (item: H.HealthItem, idx: number) => {
    const config = this.props.health.getStatusConfig();
    const isValueInConfig =
      config && this.props.health.health.statusConfig ? this.props.health.health.statusConfig.value > 0 : false;

    const showTraffic = item.children
      ? item.children.filter(sub => {
          const showItem = sub.value && sub.value > 0;
          return (sub.status !== H.HEALTHY && showItem) || !this.props.tooltip;
        }).length > 0
      : false;
    return showTraffic ? (
      <div key={idx}>
        <strong>
          {' ' + item.title + (item.text && item.text.length > 0 ? ': ' : '')}{' '}
          {config && <InfoAltIcon color={PfColors.Gray} />}
        </strong>
        {item.text}
        {item.children && (
          <ul style={{ listStyleType: 'none', paddingLeft: 12 }}>
            {item.children.map((sub, subIdx) => {
              const showItem = sub.value && sub.value > 0;
              return (sub.status !== H.HEALTHY && showItem) || !this.props.tooltip ? (
                <li key={subIdx}>
                  {createIcon(sub.status)} {sub.text}
                </li>
              ) : (
                <></>
              );
            })}
            {config && isValueInConfig && (
              <li key={'degraded_failure_config'}>
                {createIcon(H.DEGRADED)}: {config.degraded === 0 ? '>' : '>='}
                {config.degraded}% {createIcon(H.FAILURE)}: {config.degraded === 0 ? '>' : '>='}
                {config.failure}%
              </li>
            )}
          </ul>
        )}
      </div>
    ) : (
      <></>
    );
  };

  renderChildren = (item: H.HealthItem, idx: number) => {
    return item.title.startsWith(H.TRAFFICSTATUS)
      ? this.renderErrorRate(item, idx)
      : (item.status !== H.HEALTHY || !this.props.tooltip) && (
          <div key={idx}>
            <strong>{' ' + item.title + (item.text && item.text.length > 0 ? ': ' : '')}</strong>
            {item.text}
            {item.children && (
              <ul style={{ listStyleType: 'none', paddingLeft: 12 }}>
                {item.children.map((sub, subIdx) => {
                  return (
                    <li key={subIdx}>
                      {createIcon(sub.status)} {sub.text}
                    </li>
                  );
                })}
              </ul>
            )}
          </div>
        );
  };

  render() {
    const health = this.props.health;
    return health.health.items.map((item, idx) => {
      return this.renderChildren(item, idx);
    });
  }
}
