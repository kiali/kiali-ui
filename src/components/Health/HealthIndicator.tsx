import * as React from 'react';
import { PopoverPosition, Text, TextContent, TextVariants, Tooltip } from '@patternfly/react-core';
import { HealthDetails } from './HealthDetails';
import * as H from '../../types/Health';
import { createIcon } from './Helper';
import './Health.css';
import { PfColors } from '../Pf/PfColors';

export enum DisplayMode {
  LARGE,
  SMALL
}

interface Props {
  id: string;
  health?: H.Health;
  mode: DisplayMode;
  tooltipPlacement?: PopoverPosition;
}

interface HealthState {
  globalStatus: H.Status;
}

export class HealthIndicator extends React.PureComponent<Props, HealthState> {
  static getDerivedStateFromProps(props: Props) {
    return {
      globalStatus: props.health ? props.health.getGlobalStatus() : H.NA
    };
  }

  constructor(props: Props) {
    super(props);
    this.state = HealthIndicator.getDerivedStateFromProps(props);
  }

  render() {
    if (this.props.health) {
      if (this.props.mode === DisplayMode.SMALL) {
        return this.renderSmall(this.props.health);
      } else {
        return this.renderLarge(this.props.health);
      }
    }
    return <span />;
  }

  renderSmall(health: H.Health) {
    return this.renderPopover(health, createIcon(this.state.globalStatus, 'sm'));
  }

  renderLarge(health: H.Health) {
    const spanStyle: React.CSSProperties = {
      color: this.state.globalStatus.color,
      fontWeight: 'bold',
      position: 'relative',
      top: -9,
      left: 10
    };
    return (
      <>
        {createIcon(this.state.globalStatus, 'lg')}
        <span style={spanStyle}>{this.state.globalStatus.name}</span>
        <br />
        <br />
        <HealthDetails health={health} />
      </>
    );
  }
  renderHealthTooltip(health: H.Health) {
    return (
      <TextContent style={{ color: PfColors.White }}>
        <Text component={TextVariants.h2}>{this.state.globalStatus.name}</Text>
        <HealthDetails health={health} />
      </TextContent>
    );
  }

  renderPopover(health: H.Health, icon: JSX.Element) {
    return (
      <Tooltip
        aria-label={'Health indicator'}
        content={this.renderHealthTooltip(health)}
        position={PopoverPosition.auto}
        className={'health_indicator'}
      >
        {icon}
      </Tooltip>
    );
  }
}
