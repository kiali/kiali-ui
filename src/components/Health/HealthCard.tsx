import * as React from 'react';
import { Card, CardActions, CardBody, CardHead, CardHeader, Stack, StackItem, Title } from '@patternfly/react-core';
import { DisplayMode, HealthIndicator } from './HealthIndicator';
import { style } from 'typestyle';
import * as H from '../../types/Health';

type Props = {
  name: string;
  health?: H.Health;
};

const titleStyle = style({
  margin: '15px 0 11px 0'
});

class HealthCard extends React.PureComponent<Props> {
  render() {
    return (
      <Card>
        <CardHead>
          <CardActions />
          <CardHeader>
            <Title style={{ float: 'left' }} headingLevel="h3" size="2xl">
              Health
            </Title>
          </CardHeader>
        </CardHead>
        <CardBody>
          <Stack className={'stack_service_details'}>
            <StackItem id={'health'}>
              <Title headingLevel="h3" size="lg" className={titleStyle}>
                Overall Health
              </Title>
              <HealthIndicator id={this.props.name} health={this.props.health} mode={DisplayMode.LARGE} />
            </StackItem>
          </Stack>
        </CardBody>
      </Card>
    );
  }
}

export default HealthCard;
