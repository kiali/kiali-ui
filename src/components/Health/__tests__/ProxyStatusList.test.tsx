import * as React from 'react';
import { shallow } from 'enzyme';
import ProxyStatusList from '../ProxyStatusList';
import { ProxyStatus } from '../../../types/Health';
import { FlexItem, Split } from '@patternfly/react-core';
import { shallowToJson } from 'enzyme-to-json';

const workloadName = 'reviews-v1-12gaas-2as';
const shallowComponent = (statuses: ProxyStatus[]) => {
  return shallow(<ProxyStatusList workloadName={workloadName} statuses={statuses} />);
};

describe('ProxyStatusList', () => {
  describe('when status is empty', () => {
    const statuses: ProxyStatus[] = [];
    const subject = shallowComponent(statuses);

    it("doesn't render any component", () => {
      expect(subject).toHaveLength(0);
    });

    it('match the snapshot', () => {
      expect(shallowToJson(subject)).toMatchSnapshot();
    });
  });

  describe('when there are statuses', () => {
    const statuses: ProxyStatus[] = [
      { component: 'RDS', status: 'STALE' },
      { component: 'CDS', status: 'NOT_SENT' }
    ];

    const subject = shallowComponent(statuses);

    it('match the snapshot', () => {
      expect(shallowToJson(subject)).toMatchSnapshot();
    });

    it('renders an split component', () => {
      expect(subject.find(Split)).toHaveLength(1);
    });

    it('has the workload names in the first item', () => {
      const split = subject.find(Split);
      expect(split.childAt(0).html()).toContain(workloadName);
    });

    it('renders all statuses sorted alphabetically', () => {
      const split = subject.find(Split);
      expect(split.children()).toHaveLength(2);

      const statusList = split.childAt(1);
      expect(statusList).toBeDefined();

      const statusItems = statusList.find(FlexItem);
      expect(statusItems).toHaveLength(2);
      expect(statusItems.at(0).html()).toContain('CDS: NOT_SENT');
      expect(statusItems.at(1).html()).toContain('RDS: STALE');
    });
  });
});
