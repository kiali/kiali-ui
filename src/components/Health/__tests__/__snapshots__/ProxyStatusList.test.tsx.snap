// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProxyStatusList when status is empty match the snapshot 1`] = `null`;

exports[`ProxyStatusList when there are statuses match the snapshot 1`] = `
<Split
  gutter="sm"
>
  <SplitItem
    className="f44l9nu"
  >
    reviews-v1-12gaas-2as
  </SplitItem>
  <SplitItem>
    <Flex
      style={
        Object {
          "flex": "inlineFlex",
        }
      }
    >
       
      <FlexItem
        key="proxy-status-0"
      >
        <Text
          className="f1dm535y"
          component="small"
        >
          CDS: NOT_SENT
        </Text>
      </FlexItem>
      <FlexItem
        key="proxy-status-1"
      >
        <Text
          className="f1dm535y"
          component="small"
        >
          RDS: STALE
        </Text>
      </FlexItem>
       
    </Flex>
  </SplitItem>
</Split>
`;
